src/cancelled.rs

```diff
@@ -10,15 +12,25 @@ use std::panic::{self, UnwindSafe};
 #[derive(Debug)]
 #[non_exhaustive]
 pub enum Cancelled {
+    /// The query was operating but the local database execution has been cancelled.
+    Cancelled(UncancelGuard),
+
     /// The query was operating on revision R, but there is a pending write to move to revision R+1.
-    #[non_exhaustive]
     PendingWrite,
 
     /// The query was blocked on another thread, and that thread panicked.
-    #[non_exhaustive]
     PropagatedPanic,
 }
 
+#[derive(Debug)]
+pub struct UncancelGuard(pub(crate) CancellationToken);
+
+impl Drop for UncancelGuard {
+    fn drop(&mut self) {
+        self.0.uncancel();
+    }
+}
+
```

Comment:
Not too fond of this way to reset the guard. I think we can make this work in (un)attach instead